third_party/amd/lib/Analysis/RangeAnalysis.cpp

```diff
@@ -60,45 +62,142 @@ void getEnclosingLoops(Operation &op, SmallVector<LoopLikeOpInterface> &ops) {
+  auto bitWidth = mlir::ConstantIntRanges::getStorageBitwidth(elTy);
+  setResultRange(result,
+                 ConstantIntRanges::range(
+                     /*min*/ {/*numBits*/ bitWidth, /*val*/ op->getStart(),
+                              /*isSigned*/ elTy.isSigned()},
+                     /*max*/
+                     {/*numBits*/ bitWidth, /*val*/ op->getEnd(),
+                      /*isSigned*/ elTy.isSigned()},
+                     /*isSigned*/ elTy.isSigned()));
 }
 
-void inferResultRanges(tt::SplatOp *op, ArrayRef<ConstantIntRanges> argRanges,
+void inferResultRanges(tt::GatherOp *op, ArrayRef<ConstantIntRanges> argRanges,
                        SetIntRangeFn setResultRange) {
+  assert(argRanges.size() == 2 && "expected two arg ranges");
   setResultRange(op->getResult(), argRanges[0]);
 }
 
-void inferResultRanges(tt::ExpandDimsOp *op,
-                       ArrayRef<ConstantIntRanges> argRanges,
-                       SetIntRangeFn setResultRange) {
-  setResultRange(op->getResult(), argRanges[0]);
+void inferResultRangesUnaryOpForwardArgRange(
+    Operation *op, ArrayRef<ConstantIntRanges> argRanges,
+    SetIntRangeFn setResultRange) {
+  assert(op->getNumResults() == 1 && "expected op to have one result");
+  setResultRange(op->getResult(0), argRanges[0]);
+}
+
+void inferResultRangesBinaryOpUnionArgRanges(
+    Operation *op, ArrayRef<ConstantIntRanges> argRanges,
+    SetIntRangeFn setResultRange) {
+  assert(op->getNumResults() == 1 && "expected op to have one result");
+  assert(op->getNumOperands() == 2 && "expected op to have two operands");
+  assert(argRanges.size() == 2 && "expected two arg ranges");
+  setResultRange(op->getResult(0), argRanges[0].rangeUnion(argRanges[1]));
+}
+
+void inferResultRangesMaxNonNegSigned(Operation *op,
+                                      SetIntRangeFn setResultRange) {
+  for (auto result : op->getResults()) {
+    auto bitWidth =
+        mlir::ConstantIntRanges::getStorageBitwidth(result.getType());
+    setResultRange(result, ConstantIntRanges::fromSigned(
+                               APInt::getZero(bitWidth).sext(bitWidth),
+                               APInt::getMaxValue(bitWidth).sext(bitWidth)));
+  }
 }
 
 } // namespace
 
 namespace mlir::triton::AMD {
 
+bool ConstantRangeIsStaticallyTrue(const ConstantIntRanges &range, Type type) {
+  auto bitWidth = mlir::ConstantIntRanges::getStorageBitwidth(type);
+  if (bitWidth != 1)
+    return false;
+  if (auto maybeCVal = range.getConstantValue();
+      maybeCVal.value_or(APInt::getZero(bitWidth)) ==
+      APInt::getAllOnes(bitWidth))
+    return true;
+  return false;
+}
+
 void TritonIntegerRangeAnalysis::setToEntryState(
     dataflow::IntegerValueRangeLattice *lattice) {
-  propagateIfChanged(lattice, lattice->join(IntegerValueRange::getMaxRange(
-                                  lattice->getAnchor())));
+  auto anchor = lattice->getAnchor();
```

Comment:
Can you add some overall doc to explain what we are doing here? Something like scanning through users of the current value to see if there are cmp op with result assumptions. So that we can better constrain the range. Easier to follow for others with it.

Reply:
I refactored it so that it's much clearer what's going on (descriptive function names and such) and also added a doc.

Reply:
Thanks for the refactoring and great doc!